File: src/main/java/org/tindalos/principle/infrastructure/plugin/DesignQualityCheckerMojo.java
package org.tindalos.principle.infrastructure.plugin;

import java.util.Optional;

import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.tindalos.principle.domain.constraints.exception.InvalidConfigurationException;
import org.tindalos.principle.domain.resultprocessing.reporter.AnalysisResultsReporter;
import org.tindalos.principle.infrastructure.ConstraintsReader;
import org.tindalos.principle.infrastructure.di.PoorMansDIContainer;
import org.tindalos.principle.infrastructure.reporters.ReportsDirectoryManager;
import org.tindalos.principle.utils.logging.SimpleLogger;
import org.tindalos.principle.utils.logging.TheLogger;

@Mojo(name = "check")
public class DesignQualityCheckerMojo extends AbstractMojo {

    @Parameter(property = "check.location")
    private String location;

    @Override
    public void execute() throws MojoFailureException {
        TheLogger.setLogger(new SimpleLogger() {
            @Override
            public void info(String msg) {
                getLog().info(msg);
            }

            @Override
            public void error(String msg) {
                getLog().info(msg);
            }
        });

        ReportsDirectoryManager.ensureReportsDirectoryExists();

        var plan = ConstraintsReader.readFromFile(Optional.ofNullable(location));

        var printer = new LogPrinter(getLog());
        var analyzer = PoorMansDIContainer.buildAnalyzer(plan.basePackage());
        var reporter = PoorMansDIContainer.createReporter();
        try {
            var result = analyzer.analyze(plan);
            printer.printInfo(reporter.summary(result));
            if (result.hasViolations()) {
                throw new MojoFailureException("\nNumber of violations exceeds allowed limits!");
            }
        } catch (InvalidConfigurationException ex) {
            throw new MojoFailureException(ex.getMessage());
        } catch (RuntimeException ex) {
            throw new MojoFailureException("Unexpected error", ex);
        }
    }
}

File: src/main/scala/org/tindalos/principle/infrastructure/plugin/DesignQualityCheckerMojo.scala
package org.tindalos.principle.infrastructure.plugin

import java.io.IOException
import org.apache.maven.plugin.{AbstractMojo, MojoFailureException}
import org.apache.maven.plugins.annotations.{Mojo, Parameter}
import org.tindalos.principle.domain.core.AnalysisPlan
import org.tindalos.principle.domain.constraints.exception.InvalidConfigurationException
import org.tindalos.principle.infrastructure.ConstraintsReader
import org.tindalos.principle.infrastructure.di.PoorMansDIContainer
import org.tindalos.principle.infrastructure.reporters.ReportsDirectoryManager
import org.tindalos.principle.utils.logging.{SimpleLogger, TheLogger}

@Mojo(name = "check")
class DesignQualityCheckerMojo extends AbstractMojo {

  @Parameter(property = "check.location")
  private var location: String = null

  def execute(): Unit = {

    TheLogger.setLogger(new SimpleLogger() {
      override def info(msg: String) = {
        getLog().info(msg)
      }

      override def error(msg: String) = {
        getLog().info(msg)
      }
    })

    ReportsDirectoryManager.ensureReportsDirectoryExists()

    val plan = ConstraintsReader.readFromFile(java.util.Optional.ofNullable(location))

    val analyser = PoorMansDIContainer.buildAnalyzer(plan.basePackage(), new LogPrinter(getLog()))
    try {
      val result = analyser.analyze(plan)
      if (!result.success) throw new MojoFailureException("\nNumber of violations exceeds allowed limits!")
    } catch {
      case ex: IOException => getLog().error("/target/classes not found! " + ex.getMessage())
      case ex: InvalidConfigurationException => throw new MojoFailureException(ex.getMessage())
      case ex: Exception ⇒ throw new MojoFailureException("Unexpected error", ex)
    }

  }
}
